src/graph_impl/stable_graph/serialization.rs

```diff
@@ -197,30 +196,47 @@ where
         E2: Error,
     {
         let ty = PhantomData::<Ty>::from_deserialized(input.edge_property)?;
-        let mut nodes = input.nodes;
         let node_holes = input.node_holes;
         let edges = input.edges;
         if edges.len() >= <Ix as IndexType>::max().index() {
             Err(invalid_length_err::<Ix, _>("edge", edges.len()))?
         }
+        
+        let mut nodes = Vec::with_capacity(input.nodes.len() + node_holes.len());
+
+        let mut compact_nodes = input.nodes.into_iter();
+        let mut node_pos = 0;
+        for hole_pos in node_holes.iter() {
+            nodes.extend(compact_nodes.by_ref().take(hole_pos.index() - node_pos));
+            nodes.push(Node {
+                weight: None,
+                next: [EdgeIndex::end(); 2],
+            });
+            node_pos = hole_pos.index() + 1;
+            debug_assert_eq!(nodes.len(), node_pos);
```

Comment:
Debug assertions are nice but here it's perpelexing since it doesn't mesh with the aforementioned mindset that the input is untrusted and arbitrary. We don't panic on invalid input, we return an error.

Reply:
This debug assertion is really just to double-check the indices.